ci-operator multi-stage: create secret

bbguimaraes opened the pull request:
Each test is assigned a secret with a pre-defined name (already
accessible via the `$JOB_NAME_SAFE` environment variable) that will be
mounted at a specific path in all pods (from `pre` to `post`) and can be
used to pass a small amount of data to subsequent steps.

bbguimaraes commented on pkg/steps/multi_stage.go:
```diff
@@ -20,6 +20,7 @@ import (
 
 const (
 	multiStageTestLabel = "ci.openshift.io/multi-stage-test"
+	secretMountPath     = "/var/run/secrets/ci.openshift.io/multi-stage"
```
/shrug

bbguimaraes commented:
> If someone writes data into that directory, how do we backfill that data into the secret? Will this be easier than using a persistent volume?

We don't. One of the design decisions for E2E tests was that test steps would not need to share data, all communication should be done through the test cluster.

The secret in that case is only used to expose the `kubeconfig` that grants access to the test cluster and to store data needed to tear it down. Only one step in the test modifies the secret, it uses a single `curl` command to do so (it can be done with `oc`, if available), and it is likely to be written once and shared via references or workflows.

Because of that, I don't think persistent volumes are necessary. We could make it easier to transfer data from the container to the secret (as previously discussed), but it doesn't seem worth it. It can be added later if we determine it is necessary.

stevekuznetsov commented:
@bbguimaraes you mean `curl` in the actual job configuration? That should not be in the domain of the test. We should consider a sidecar or something to do the upload for people. I don't think that users should be using the API using `curl` in their test to get this functionality

bbguimaraes commented:
> @bbguimaraes you mean `curl` in the actual job configuration?

Yes, the last command in (my version of) the `installer` step is a PUT to the API server to update the secret.

> That should not be in the domain of the test. We should consider a sidecar or something to do the upload for people. I don't think that users should be using the API using `curl` in their test to get this functionality

A sidecar would require either equally intrusive changes to the test or cooperation from `ci-operator` similar to what is done with artifacts (*shudder*). A wrapper for the test command, as you've suggested in the past, would be better suited. Given that it would be completely independent of this PR, which is only concerned with creating the secret and adding the mounts, can we agree to merge it and add a separate task to create and use that tool?

stevekuznetsov commented:
Why not pull the data back in `ci-operator` and create the secret from there, a la pushing config up to GCS?

bbguimaraes commented:
As decided in private discussion, the secret should be updated automatically from the contents of a directory.

Rebased, added a tool that does that, integrated it with multi-stage tests, tested the E2E suite. The PR is now `XL`, but https://github.com/openshift/ci-tools/pull/171/commits/70bfb69535bb3e06929559377bcd281cddd9026c is the rebased version of the old commit and half the other changes are unit and integration test updates.

stevekuznetsov commented on cmd/secret-wrapper/main.go:
```diff
@@ -0,0 +1,94 @@
+)
+
+func main() {
+	if err := run(os.Args); err != nil {
```
Why not use `entrypoint`?

bbguimaraes replied:
There isn't much overlap other than the very simple function that executes a program, and we would still need a separate container to create the secret when `entrypoint` finishes.

stevekuznetsov commented on cmd/secret-wrapper/main.go:
```diff
@@ -0,0 +1,94 @@
+)
+
+const (
+	DIR = "/tmp/secret"
```
nit: this is not idomatic Go

bbguimaraes replied:
I'll admit it took me a long time to realize the capitalization was the problem (had to `grep` for `const`) =)

bbguimaraes commented on cmd/secret-wrapper/main.go:
```diff
@@ -0,0 +1,94 @@
+		return err
+	}
+	if err := execCmd(os.Args); err != nil {
+		return fmt.Errorf("failed to execute wrapped command: %v\n", err)
```
Bad refactoring.

bbguimaraes commented:
> I worry there are on-exit bugs we've solved in entrypoint that will need to be re-solved since we're not reusing it

I completely understand. I had a look at the `git` history when evaluating `entrypoint`, problems were usually either in the interaction with other components, the timeout, or the files it creates, none of which are required in our case — it could be argued those bugs are a reason *not* to use it.

`secret-wrapper` follows kube's [documentation](https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods) and should be sufficient for the narrower case it supports.

bbguimaraes commented:
Added integration tests.

stevekuznetsov commented:
/lgtm

stevekuznetsov commented:
/hold cancel
(oops)
